File: LeetCode/2779 Maximum Beauty of an Array After Applying Operation.py
from typing import List


class Solution:
    def maximumBeauty(self, nums: List[int], k: int) -> int:
        max_num = max(nums)
        memo = [0] * (max_num + k + 10)
        for num in nums:
            memo[max(0, num - k)] += 1
            memo[num + k + 1] -= 1
        for i in range(1, len(memo)):
            memo[i] += memo[i - 1]
        return max(memo)

File: LeetCode/2779 Maximum Beauty of an Array After Applying Operation.kt
class Solution {
    fun maximumBeauty(nums: IntArray, k: Int): Int {
        val maxVal = nums.max()
        val numMemo = ArrayList(List(maxVal+k+10) { 0 })
        for (num in nums) {
            numMemo[maxOf(0, num-k)] += 1
            numMemo[num+k+1] -= 1
        }
        for (i in 1 until numMemo.size) {
            numMemo[i] += numMemo[i-1]
        }
        return numMemo.max()
    }
}